feat!: single use claims

mverzilli opened the pull request:
Adds a state var type: `SingleUseClaim`, which is a useful primitive to build semaphore protocols (see [#1590](https://github.com/AztecProtocol/aztec-packages/issues/15901) for a detailed discussion of semaphores and alternative implementations of this primitive).

It also ports the `PrivateVoting` contract example to rely on it, which shows that the primitive helps abstract contract devs from low-level details. 

Closes F-132
Closes https://github.com/AztecProtocol/aztec-packages/issues/15901

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/single_use_claim.nr:
```diff
@@ -0,0 +1,266 @@
+///
+/// A private state variable type that represents a right a user can exercise at most once.
+///
+/// You can declare a `SingleUseClaim` state variable within your contract's `#[storage]` struct:
```
nit: I'd have the second paragraph instead show examples of the kinds of things you can use this for, and then go into usage etc. It'd make `ClaimIdentifier` more obvious from the get-go.

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/single_use_claim.nr:
```diff
@@ -0,0 +1,266 @@
+///
+/// #### Utility functions
+/// - `is_done(claim_id: ClaimIdentifier) -> bool`: returns whether the owner has exercised the given claim.
+/// Checks existence of a nullifier.
```
This is not needed because the docsite will automatically pull these by parsing the source code. These explanations would go in the docs for each.

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/single_use_claim.nr:
```diff
@@ -0,0 +1,266 @@
+///     // Now that we checked that the voter is entitled to vote,
+///     // we can enqueue the corresponding public effect
+///     self.enqueue_self.add_to_tally_public(candidate);
+/// }
```
We should have some more content here, we need to include 1) cost (emitting a nullifier) and 2) privacy (fully private due to the nsk)

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/single_use_claim.nr:
```diff
@@ -0,0 +1,266 @@
+    /// SingleUseClaim state variable scoped to the given owner address.
+    ///
+    /// The SingleUseClaim state can then be written by calling `claim` on it, or read
+    /// via `has_claimed` (utility function), or `assert_claimed` (private function).
```
```suggestion
    /// The SingleUseClaim state can then be written by calling [claim] on it, or read
    /// via [has_claimed] (in utility functions), or [assert_claimed] (in private functions).
```

You can use [this] to have `nargo doc` automatically link to the corresponding function/type/etc. The LSP will also highlight the link and support ctrl+click.

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/single_use_claim.nr:
````diff
@@ -0,0 +1,266 @@
+    ///
+    /// ```noir
+    /// // Continue execution only if owner hasn't acted upon this variable before
+    /// self.storage.some_single_use_claim.at(owner).do(...);
````
`claim` instead of `do` no?

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/single_use_claim.nr:
```diff
@@ -0,0 +1,266 @@
+    ///
+    /// * `SingleUseClaimWithOwner` - The original `SingleUseClaimWithOwner` augmented with an `owner` attribute.
+    ///         You can then call methods like `.claim()`, `.assert_claimed()`, `.has_claimed()`, etc. which
+    ///         are scoped to the given owner address.
```
We don't need to follow a rigid structure with `Arguments` and `Returns` sections, e.g. `owner` in this case has already been described.

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/single_use_claim.nr:
```diff
@@ -0,0 +1,266 @@
+    /// `owner` claimed the given `claim_id`.
+    pub fn assert_claim(self, claim_id: ClaimIdentifier) {
+        let owner_nsk_app = self.context.request_nsk_app(get_public_keys(self.owner).npk_m.hash());
+        let nullifier = self.compute_nullifier(claim_id, owner_nsk_app);
```
Here is where we'd do `check_nullifier_exists`. Possibly might move into `push_null_read_reqs`, not sure.

mverzilli replied:
I see what you mean, this unsafe-check/assert-result incantation is likely to repeat almost anywhere we need to do sth like this. 

On the other hand, moving it into `push_null_read_reqs` means anyone using that will pay the overhead of this oracle call... maybe we can wait and see if the need arises a second or third time before lifting it

nventuro commented on noir-projects/noir-contracts/contracts/app/private_voting_contract/src/main.nr:
```diff
@@ -1,27 +1,17 @@
-    use dep::aztec::state_vars::{Map, PublicImmutable, PublicMutable};
+
+    #[derive(Serialize)]
+    struct Vote {}
```
Tried writing a suggestion for docs for this, explaining why the struct is empty, and ended up thinking that we might instead use `Map`. E.g. if there are multiple elections, instead of 

```noir
struct Election {
  id: Field,
}

fn vote(election_id, candidate) {
  self.storage.claims.at(owner).claim(Election { election_id });
}
```

we'd do

```noir
storage {
  election_claims: Map<ElectionId, Claims>,
}

fn vote(election_id, candidate) {
  self.storage.election_claims.at(election_id).at(owner).claim();
}
```

and then get rid of `CLAIM` altogether. wdyt?

mverzilli replied:
Trying to think a bit ahead of how this would compose with @benesjan 's work on `Owned`. Would the end game of this be:

```noir
storage {
  election_claims: Map<ElectionId, Owned<Claims>>,
}

fn vote(election_id, candidate) {
  self.storage.election_claims.at(election_id).at(owner).claim();
}
```

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/single_use_claim/test.nr:
```diff
@@ -0,0 +1,173 @@
+    let owner = env.create_light_account();
+
+    let claim_id = MockStruct::new(1, 2);
+    expect_not_yet_claimed(env, STORAGE_SLOT, claim_id, owner);
```
This helper is so simple that I'm not 100% sold on it being better than just inlining the contents - I ended up going back and forth to the definition to remind myself what it is the helper did exactly. Up to you though.

mverzilli replied:
You're right, I think it started out being more complex but then it got dumbed down to the point that its existence is no longer justified

nventuro reviewed (commented):
Very very nice, thanks! Left some suggestions re. docs and the `CLAIMS` generic type.

mverzilli commented on barretenberg/acir_tests/yarn.lock:
```diff
@@ -1167,13 +1167,6 @@ __metadata:
   languageName: node
   linkType: hard
 
-"base64-js@npm:^1.3.1":
```
mmm, I find it odd that I'm causing a yarn.lock to change. will revert this one.

iAmMichaelConnor commented on noir-projects/aztec-nr/aztec/src/state_vars/single_use_claim.nr:
```diff
@@ -0,0 +1,148 @@
+/// this leaks the fact that _someone_ exercised that claim once, but the owner is hidden
+/// behind the owner's app-siloed nullifier secret key (aka `nsk`).
+///
+/// Public effects you emit alongside a claim (e.g. updating a tally) may still let observers infer
```
```suggestion
/// Public effects you emit alongside a claim (e.g. a public function call to update a public tally) may still let observers infer
```

iAmMichaelConnor commented on noir-projects/aztec-nr/aztec/src/state_vars/single_use_claim.nr:
```diff
@@ -0,0 +1,148 @@
+/// ## Privacy considerations
+///
+/// The [claim] function emits a public nullifier per `owner`:
+/// this leaks the fact that _someone_ exercised that claim once, but the owner is hidden
```
> this leaks the fact that _someone_ exercised that claim once

Technically, the act of emitting the nullifier doesn't leak that a claim was executed (or even that this contract was executed). It's the other app-specific logic (e.g. the public effects that you allude to a few lines later) that might leak the fact that someone exercised a claim.

nventuro replied:
Agreed, this is incorrect. It should say that there is no leak as the owner's nsk hides the nullifier.

iAmMichaelConnor commented on noir-projects/aztec-nr/aztec/src/state_vars/single_use_claim.nr:
```diff
@@ -0,0 +1,148 @@
+    /// _eventually_ cause the current transaction simulation/proving to fail if the nullifier
+    /// is not found, which guarantees any private function using this call will fail unless
+    /// `owner` has already claimed this single use claim.
+    pub fn assert_claim(self) {
```
Just a suggestion:

```suggestion
    pub fn assert_claimed(self) {
```

iAmMichaelConnor commented on noir-projects/noir-contracts/contracts/app/private_voting_contract/src/main.nr:
```diff
@@ -1,80 +1,95 @@
+        // election => election started at
+        active_at_block: Map<ElectionId, PublicImmutable<u32, Context>, Context>,
+        // election => voter => single use claim that ensures voter can at most vote once per election
+        vote_claims: Map<ElectionId, Owned<SingleUseClaim<Context>, Context>, Context>,
```
Q: should the word `Private` appear somewhere in this type? `OwnedPrivateState`? `SingleUsePrivateClaim`? I'm wondering if it's clear that this is a _private_ state variable to newcomers.

nventuro replied:
Ehhhh, I can see where you're coming from, but it seems a bit verbose. Idk. @ciaranightingale?

mverzilli replied:
Fair point. The trade off is that these types declarations are getting super verbose, not sure on which side to err

mverzilli replied:
If we want to add `Private` somewhere, I think I would go with `PrivateSingleUseClaim` (and then we make a convention out of "all scalar-ish state var types start with either `Private` or `Public`). It might sound less proper English, but it's easier to parse IMO 

@benesjan @nventuro?

iAmMichaelConnor commented on noir-projects/noir-contracts/contracts/app/private_voting_contract/src/main.nr:
```diff
@@ -1,80 +1,95 @@
+        // admin can start and end elections
+        admin: PublicMutable<AztecAddress, Context>,
+        // election => candidate => number of votes
+        tally: Map<ElectionId, Map<Field, PublicMutable<Field, Context>, Context>, Context>,
```
Only half joking -- and obvs a discussion that shouldn't be on this PR: Maybe we should replace `Context` with `Ω` because it's sooo much less verbose, and it's awesome. And Ω meaning "everything" fits quite well the the Context which basically does everything. (alt + z on my macbook)

nventuro replied:
Ω more typically means 'the end'. Also, wtf.

mverzilli replied:
My nerd brain is really eager to go full on the Ω idea, but Aztec already has so many novel concepts to grok that adding a Greek character to the mix feels like spending from an attention budget that is already tight 😅.

But the type variable doesn't need to be called Context. We could call it `C` and at least make this look a bit more succinct.

Eventually it would be nice to get rid if it altogether. I suspect the people from Haskell or Idris have some neat solution to this problem

iAmMichaelConnor replied:
I like the idea of `C`!

iAmMichaelConnor reviewed (commented):
Looks awesome! Only pedantic comments from me.
I'm really happy to see this come into existence!

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/single_use_claim.nr:
```diff
@@ -0,0 +1,148 @@
+
+mod test;
+
+/// A private state variable type that represents a right a user can exercise at most once.
```
```suggestion
/// A private state variable type that represents a right each user can exercise at most once.
```

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/single_use_claim.nr:
```diff
@@ -0,0 +1,148 @@
+mod test;
+
+/// A private state variable type that represents a right a user can exercise at most once.
+///
```
This should mention that this is an `OwnedStateVariable` and as such needs to be inside the `Owned` container. See how `Owned` pops up out of the blue in the example here.

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/single_use_claim.nr:
```diff
@@ -0,0 +1,148 @@
+///     let voter = self.msg_sender().unwrap();
+///
+///     // `claim` will only succeed if `voter` hasn't ever run this before.
+///     self.storage.vote_claims.at(election_id).at(voter).claim();
```
```suggestion
///     // `claim` will only succeed if `voter` hasn't already claimed for this `election_id`.
///     self.storage.vote_claims.at(election_id).at(voter).claim();
```

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/single_use_claim.nr:
```diff
@@ -0,0 +1,148 @@
+///
+/// [SingleUseClaim] state variables are declared within a contract's `#[storage]` struct:
+///
+/// `your_claimable: SingleUseClaim<Context>`
```
This is wrong, no? `Owned` is missing

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/single_use_claim.nr:
```diff
@@ -0,0 +1,148 @@
+///
+/// ## Cost
+///
+/// Exercising a claim (via the [claim] function) emits one nullifier.
```
```suggestion
/// Exercising a claim (via the [SingleUseClaim::claim] function) emits one nullifier.
```

Don't you need to do this to get the links to work? I think I do on my machine.

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/single_use_claim.nr:
```diff
@@ -0,0 +1,148 @@
+    ///
+    /// Any subsequent attempts at calling this function with the same owner
+    /// are guaranteed by the protocol to fail (since it would result in duplicate
+    /// nullifiers), which is why this function provides `at most once execution` semantics.
```
```suggestion
```

Saying 'the fn emits a nullif' is enough to convey the point.

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/single_use_claim.nr:
```diff
@@ -0,0 +1,148 @@
+impl SingleUseClaim<&mut PrivateContext> {
+    /// Records the fact that the owner in scope has claimed this single use claim. This function succeeds at most once
+    /// for a given `owner`. Any subsequent calls with same owner are guaranteed to fail.
+    ///
```
```suggestion
    /// Note that, by itself, [claim] doesn't do anything other than prevent [claim] from being called again
    /// with the same owner. Its invocation is typically accompanied by other contract actions, e.g. minting
    /// tokens, casting a vote, etc.
    ///
```

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/single_use_claim.nr:
```diff
@@ -0,0 +1,148 @@
+        self.context.push_nullifier(self.compute_nullifier(owner_nsk_app));
+    }
+
+    /// Asserts that the owner has claimed this single use claim.
```
```suggestion
    /// Asserts that the owner has already claimed this single use claim, i.e. that [claim] has been called.
```

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/single_use_claim.nr:
```diff
@@ -0,0 +1,148 @@
+    /// during the current transaction (also known as _transient nullifiers_). This request will
+    /// _eventually_ cause the current transaction simulation/proving to fail if the nullifier
+    /// is not found, which guarantees any private function using this call will fail unless
+    /// `owner` has already claimed this single use claim.
```
```suggestion
    /// This function generates a kernel nullifier read request, so it can even assert the existence of
    /// pending claims, i.e. when [claim] was called in the same transaction as [assert_claim].
```

I don't think we need to go into too much detail re. how read requests work, expectations around those, etc., the fact that we create one is enough. Highlighting that pending claims can be read might be useful in a cursory glance, as it is very relevant information for the user.

Also, a note on names: 'transient' notes are those that only existed for some time in a transaction and were then squashed and not persisted. A note or nullifier that has not yet been inserted into the tree (i.e. one from the current tx) is called 'pending' instead. The opposite of 'pending' is 'settled'. Only pending notes can become transient via squashing.

('pending' nullifiers exist, but there are no 'transient' nullifiers).

nventuro commented on noir-projects/noir-contracts/contracts/app/private_voting_contract/src/test/first.nr:
```diff
@@ -39,56 +44,121 @@ unconstrained fn test_end_vote() {
+
+    env.public_context_at(voting_contract_address, |context| {
+        let election_42_storage_slot =
+            derive_storage_slot_in_map(PrivateVoting::storage_layout().tally.slot, election_id_42);
```
Better to have the contract have a getter for this (the tally for an id i mean)

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/single_use_claim.nr:
```diff
@@ -0,0 +1,147 @@
+    /// the `owner`'s nullifier secret key and the state variable's storage slot.
+    ///
+    /// It then requests PXE to find the nullifier in private state, including nullifiers emitted
+    /// during the current transaction (also known as _transient nullifiers_).
```
I'd remove this bit.

nventuro reviewed (commented):
(forgot I had some review comments unsubmitted to posting this)

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/single_use_claim.nr:
```diff
@@ -0,0 +1,137 @@
+
+/// A private state variable type that represents a right each user can exercise at most once. Because it is
+/// _owned_, you must wrap a [SingleUseClaim] inside an [crate::state_vars::owned::Owned] state variable when
+/// using it in storage:
```
```suggestion
/// A private state variable type that represents a right each user can exercise at most once.
///
/// Because it is _owned_, you must wrap a [SingleUseClaim] inside an [crate::state_vars::owned::Owned]
/// state variable when using it in storage:
```

We try to make it so that the first paragraph is a single short sentence that describes the thing simply.

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/single_use_claim.nr:
```diff
@@ -0,0 +1,137 @@
+    ///
+    /// The implementation of this function emits a nullifier derived from
+    /// the `owner`'s nullifier secret key and the storage slot of the underlying state variable.
+    pub fn claim(self) {
```
```suggestion
    /// Records the fact that the owner in scope has claimed this single use claim.
    ///
    /// This function succeeds at most once for a given `owner`. Any subsequent calls with same owner are
    /// guaranteed to fail.
    ///
    /// Note that, by itself, calling this function doesn't do anything other than prevent it being called
    /// again for the same owner. Its invocation is typically accompanied by other contract actions,
    /// e.g. minting tokens, casting a vote, etc.
```

Same comment re. first line, and also removed links to self (which seem more confusing that helpful) and unecessary (?) adv comments.

nventuro reviewed (approved):
Fantastic, great to see this go in!

mverzilli committed:
feat!: single use claims

Adds a state var type: `SingleUseClaim`, which is a useful primitive to build semaphore protocols (see [#1590](https://github.com/AztecProtocol/aztec-packages/issues/15901) for a detailed discussion of semaphores and alternative implementations of this primitive).

It also ports the `PrivateVoting` contract example to rely on it, which shows that the primitive helps abstract contract devs from low-level details.

Closes F-132
Closes https://github.com/AztecProtocol/aztec-packages/issues/15901
